lib/languages/js/source-code/source-code.js

```diff
@@ -1164,7 +1186,11 @@ class SourceCode extends TokenStore {
      */
     finalize() {
 
-        // Step 1: ensure that all of the necessary variables are up to date
+        // We only need to do this for ESTree-compatible ASTs
+        if (!this.isESTree) {
+            return;
+        }
```

Comment:
This might be a breaking change because this code was executed for non-estree ASTs.

Reply:
Interesting. Without this, several tests failed, so I thought we were skipping logic. I'll take another look.

Reply:
Ah good call. I think it makes sense to crash in this instance because it's not actually a parsing error. The parsing succeeded but for some reason creating the `SourceCode` instance failed. I'd expect this to be caught by plugin developers before they release and most end-users would not end up seeing thing. What do you think?

Reply:
> I think it makes sense to crash in this instance because it's not actually a parsing error.

Makes sense to me 👍

But, related to this, I think we should update the failing tests to return valid parse results instead of updating parts of the code to avoid crashing on those tests (https://github.com/eslint/eslint/pull/18448#discussion_r1620549779 & https://github.com/eslint/eslint/pull/18448#discussion_r1620551343).

Reply:
Agreed.